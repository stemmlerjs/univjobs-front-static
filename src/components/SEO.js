import path from 'path';
import React from 'react';
import Helmet from 'react-helmet';
import PropTypes from 'prop-types';
import * as config from '../config';

const getSchemaOrgJSONLD = ({
  isBlogPost,
  url,
  title,
  image,
  description,
  datePublished,
}) => {
  const schemaOrgJSONLD = [
    {
      '@context': 'https://univjobs.ca',
      '@type': 'website',
      url,
      name: title,
      alternateName: config.title,
      logo: "https://univjobs.ca/static/u-logo.4bc69dc3.png"
    },
  ];

  return isBlogPost
    ? [
        ...schemaOrgJSONLD,
        {
          '@context': 'https://univjobs.ca',
          '@type': 'BreadcrumbList',
          itemListElement: [
            {
              '@type': 'ListItem',
              position: 1,
              item: {
                '@id': url,
                name: title,
                image,
              },
            },
          ],
        },
        {
          '@context': 'https://univjobs.ca',
          '@type': 'BlogPosting',
          url,
          name: title,
          alternateName: config.title,
          headline: title,
          image: {
            '@type': 'ImageObject',
            url: image,
          },
          description,
          author: {
            '@type': 'Person',
            name: 'Charles Javelona',
          },
          publisher: {
            '@type': 'Organization',
            url: 'https://univjobs.ca',
            logo: config.logo,
            name: 'Charles Javelona',
          },
          mainEntityOfPage: {
            '@type': 'WebSite',
            '@id': config.url,
          },
          datePublished,
        },
      ]
    : schemaOrgJSONLD;
};

const SEO = ({ postData, postImage, isBlogPost }) => {
  const postMeta = postData || {};

  const title = postMeta.title || config.title;
  const description =
    postMeta.description || postData.excerpt || config.description;
  const image = `${config.url}${postImage}` || config.image;
  const url = postMeta.slug
    ? `${config.url}${postMeta.slug}`
    : config.url;
  const datePublished = isBlogPost ? postMeta.date : false;

  const schemaOrgJSONLD = getSchemaOrgJSONLD({
    isBlogPost,
    url,
    title,
    image,
    description,
    datePublished,
  });
  
  return (
    <Helmet>
      {/* General tags */}
      <meta name="description" content={description} />
      <meta name="image" content={image} />

      {/* Schema.org tags */}
      <script type="application/ld+json">
        {JSON.stringify(schemaOrgJSONLD)}
      </script>

      {/* OpenGraph tags */}
      <meta property="og:url" content={url} />
      {isBlogPost ? <meta property="og:type" content="article" /> : null}
      <meta property="og:title" content={title} />
      <meta property="og:description" content={description} />
      <meta property="og:image" content={image} />

      {/* Twitter Card tags */}
      <meta name="twitter:card" content="summary_large_image" />
      <meta name="twitter:creator" content={config.twitter} />
      <meta name="twitter:title" content={title} />
      <meta name="twitter:description" content={description} />
      <meta name="twitter:image" content={image} />
    </Helmet>
  );
};

SEO.propTypes = {
  isBlogPost: PropTypes.bool,
  postData: PropTypes.shape({
    title: PropTypes.string,
    description: PropTypes.string,
    frontmatter: PropTypes.any,
    excerpt: PropTypes.any,
  }).isRequired,
  postImage: PropTypes.string,
};

SEO.defaultProps = {
  isBlogPost: false,
  postImage: "img/u.png",
};

export default SEO;