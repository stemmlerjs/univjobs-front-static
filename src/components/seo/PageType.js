const PageType = {
  REGULAR: 'REGULAR', // landing page, /blog, /categories
  JOB: 'JOB',
  CITY: 'CITY', //NOTE: Cities may have different images or keywords 
  BLOG_POST: 'BLOG_POST',
  COMPANY: 'COMPANY'
}


/**
 * TODO: 
 * - company.js
 *   - JSON-LD format
 *   - Twitter format
 *   - OG format
 * - city.js
 *   - JSON-LD format
 *   - Twitter format
 *   - OG Format
 * - team.js
 *   - JSON-LD format
 *   - Twitter format
 *   - OG format
 */
export default PageType

