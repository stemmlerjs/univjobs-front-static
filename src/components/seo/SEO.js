import path from 'path'
import React from 'react'
import Helmet from 'react-helmet'
import PropTypes from 'prop-types'
import JsonLd from './JsonLd'
import config from '../../config';

const returnBaseUrl = (url) => {
  const t = new URL(url)
  return `${t.origin}${t.pathname}`
}  


const SEO = ({ requiredProps, type, pageProps }) => (
  <Helmet>
    <title>{requiredProps.title}</title>
    <meta name="description" content={requiredProps.description} />
    <meta name="image" content={requiredProps.image} />

    <script type="application/ld+json">
    {
       JSON.stringify(JsonLd(type, requiredProps, pageProps))
    }
    </script>

    {/* OpenGraph tags 
      * TODO: 
      * - Create a module with returnable objects mapping to different OG tags such as articles, etc..
      *   Refer to http://ogp.me/
      *   For now we will do a basic OG tag. 
      * 
    */}
     <meta property="og:url" content={requiredProps.url} />
     <meta property="og:title" content={requiredProps.title} />
     <meta property="og:description" content={requiredProps.description} />
     <meta name="image" content={requiredProps.image} />

     {/* Twitter Card tags */}
       <meta name="twitter:card" content="summary_large_image" />
       <meta name="twitter:creator" content={config.twitter} />
       <meta name="twitter:site" content={config.twitter} />
       <meta name="twitter:title" content={requiredProps.title} />
       <meta name="twitter:description" content={requiredProps.description} />
       <meta name="twitter:image" content={requiredProps.image} /> 

      {/**Canonical Tags
        TODO: 
          - Monitor Google Search Console for the next few months. Adjust accordingly.
          - Add mobile variant at some point.
            Refer to <link rel="alternate" media="only screen and (max-width: 640px)"  href="http://m.example.com/dresses/green-dresses">
      */}
       <link rel="canonical" href={returnBaseUrl(requiredProps.url)} />


  </Helmet>
)

export default SEO;

SEO.propTypes = {
  requiredProps: PropTypes.shape({
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
    image: PropTypes.string.isRequired
  }).isRequired,
  type: PropTypes.string.isRequired,
  pageProps: PropTypes.object.isRequired
}
