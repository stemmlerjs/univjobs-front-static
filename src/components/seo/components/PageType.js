const PageType = {
  REGULAR: 'REGULAR', // landing page, /blog, /categories
  JOB: 'JOB',
  CITY: 'CITY', //NOTE: Cities may have different images or keywords 
  BLOG_POST: 'BLOG_POST',
  COMPANY: 'COMPANY',
  PERSON: 'PERSON'
}


/**
 * TODO: 
 * - company.js
 *   - JSON-LD format
 *   - Twitter format
 *   - OG format
 * - city.js
 *   - JSON-LD format
 *   - Twitter format
 *   - OG Format
 * - team.js
 *   - JSON-LD format
 *   - Twitter format
 *   - OG format
 * - jobs
 *   - JSON-LD format
 *   - Twitter format
 *   - OG format
 * NOTE: We will only do basic formats for now. We had broken JSON-LD and meta tags not working properly
 */

export default PageType;
