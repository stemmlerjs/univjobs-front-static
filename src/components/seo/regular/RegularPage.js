
// import Joi from 'joi-browser'
import { BasePage } from '../shared/BasePage'

export class RegularPage extends BasePage {
  constructor (title, description) {
    super(title, description);
  }
}

export function createRegularPage (title, description) {
  // Validate props here if we get any.
  return new RegularPage(title, description);
}


