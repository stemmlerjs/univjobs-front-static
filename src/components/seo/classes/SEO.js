import React from 'react'
import PropTypes from 'prop-types'
import Joi from 'joi-browser'

import PageType from '../components/PageType'

const requiredPropsSchema = Joi.object().keys({
  title: Joi.string().min(3).max(300).required(),
  description: Joi.string().min(3).max(600).required()
})

const regularPagePropsSchema = Joi.object().keys({

})

const blogPostPagePropsSchema = Joi.object().keys({
  datePublished: Joi.date().required(),
  authorName: Joi.string().min(2).max(100).required(),
  image: Joi.im

  
})

class SEO extends React.Component {
  constructor (props) {
    super(props);
  }

  validateRequiredProps () {
    const { requiredProps } = this.props;
    const result = Joi.validate(requiredProps, requiredPropsSchema);
    if (result.error) {
      console.error(result.error);
    }
  }

  validatePageProps () {
    // if require page props aren't passed in, throw an error
    const { pageType } = this.props;
    const result = Joi.validate(pageType);

  }

  render() {
    return (
      <div></div>
    )
  }
}
export default SEO;

SEO.propTypes = {
  requiredProps: PropTypes.shape({
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired
  }).isRequired,
  pageType: PropTypes.string.isRequired,
  pageProps: PropTypes.object.isRequired
}

/**
 * 
 * 
 * <SEO
  required={{
    title: 'title'
    description: 'desc'
  }}
  type={PageTypes.REGULAR}
  pageProps={{

  }}
/>
 */
