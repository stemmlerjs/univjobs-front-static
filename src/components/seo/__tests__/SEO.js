import { log } from "util";

import React from "react"
import renderer from 'react-test-renderer'
import { shallow, mount, render } from 'enzyme';
import SEO from '../SEO'
import { createRegularPage } from '../regular/RegularPage'
import PageType from '../PageType'

let regularPage;
let tree;
let componentInstance;


describe("SEO Component", () => {
  describe('Regular Page', () => {

    it("Renders using Enzyme", () => {
      regularPage = createRegularPage(
        'Univjobs | Simplifying Hiring Students', 
        "Hello friends"
      );
      const wrapper = shallow(
      <SEO 
        requiredProps={regularPage.getRequiredProps()}
        type={PageType.REGULAR}
        pageProps={regularPage.getProps()}
      />);

      console.log(wrapper.find('title').text())
    })


    it("Renders without error", () => {
      regularPage = createRegularPage(
        'Univjobs | Simplifying Hiring Students', 
        "Hello friends"
      );

      tree = renderer
        .create(
          <SEO 
            requiredProps={regularPage.getRequiredProps()}
            type={PageType.REGULAR}
            pageProps={regularPage.getProps()}
          />
        )
        .toJSON()
      expect(tree).toMatchSnapshot();
    })

    it('Has the title in the meta tags', () => {
      regularPage = createRegularPage(
        'Univjobs | Simplifying Hiring Students', 
        "Hello friends"
      );

      tree = renderer.create(
        <SEO 
          requiredProps={regularPage.getRequiredProps()}
          type={PageType.REGULAR}
          pageProps={regularPage.getProps()}
        />
      )

      // console.log(tree.root)

      // componentInstance = tree.root;
      // expect(componentInstance.findByType('title')).toEqual('hello')
    })
  })
})
