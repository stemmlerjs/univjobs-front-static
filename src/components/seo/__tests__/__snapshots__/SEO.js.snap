// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `<div />`;

exports[`SEO Component Can render a regular page 1`] = `null`;

exports[`SEO Component Can render a regular page SEO component 1`] = `null`;

exports[`SEO Component Regular Page Can render a regular page 1`] = `null`;

exports[`SEO Component Regular Page Renders without error 1`] = `null`;

exports[`SEO Component renders correctly 1`] = `null`;
